cpp/src/join/unique_hash_join.cu

```diff
@@ -0,0 +1,334 @@
+                             cudf::size_type* build_indices,
+                             cudf::size_type* probe_indices)
+{
+  auto i       = group.thread_rank();
+  auto const n = *block_counter;
+  group.sync();
+
+  __shared__ cudf::size_type offset;
+
+  if (i == 0) { offset = atomicAdd(counter, n); }
+  group.sync();
+
+  while (i < n) {
+    *(build_indices + offset + i) = buffer[i].first;
+    *(probe_indices + offset + i) = buffer[i].second;
+    i += group.size();
+  }
+  if (group.thread_rank() == 0) { *block_counter = 0; }
+  group.sync();
+}
+
+template <typename Iter, typename HashTable>
+CUDF_KERNEL void unique_join_probe_kernel(Iter iter,
+                                          cudf::size_type size,
+                                          HashTable hash_table,
+                                          cudf::size_type* build_indices,
+                                          cudf::size_type* probe_indices,
+                                          cudf::size_type* counter)
+{
+  namespace cg = cooperative_groups;
+
+  auto constexpr cg_size = HashTable::cg_size;
+
+  __shared__ cuco::pair<size_type, size_type> block_buffer[DEFAULT_JOIN_BLOCK_SIZE];
+  __shared__ size_type block_counter;
+
+  auto idx               = cudf::detail::grid_1d::global_thread_id() / cg_size;
+  auto const stride      = cudf::detail::grid_1d::grid_stride() / cg_size;
+  auto const block       = cg::this_thread_block();
+  auto const tile        = cg::tiled_partition<cg_size>(block);
+  auto const thread_rank = block.thread_rank();
+
+  if (thread_rank == 0) { block_counter = 0; }
+  block.sync();
+
+  while ((idx - stride) < size) {
+    if (idx < size) {
+      auto const found = hash_table.find(tile, *(iter + idx));
+      if (tile.thread_rank() == 0 and found != hash_table.end()) {
+        auto const offset    = atomicAdd_block(&block_counter, 1);
+        block_buffer[offset] = cuco::pair{static_cast<cudf::size_type>(found->second),
+                                          static_cast<cudf::size_type>(idx)};
+      }
+    }
+    block.sync();
+    if (block_counter > (DEFAULT_JOIN_BLOCK_SIZE / 2)) {
+      flush_buffer(block, &block_counter, block_buffer, counter, build_indices, probe_indices);
+    }
+    idx += stride;
+  }
+  block.sync();
+  if (block_counter > 0) {
+    flush_buffer(block, &block_counter, block_buffer, counter, build_indices, probe_indices);
+  }
+}
+}  // namespace
+
+template <typename Hasher, cudf::has_nested HasNested>
+unique_hash_join<Hasher, HasNested>::unique_hash_join(cudf::table_view const& build,
+                                                      cudf::table_view const& probe,
+                                                      bool has_nulls,
+                                                      cudf::null_equality compare_nulls,
+                                                      rmm::cuda_stream_view stream)
+  : _has_nulls{has_nulls},
+    _is_empty{build.num_rows() == 0},
```

Comment:
Should this be `build.num_rows() == 0 || probe.num_rows() == 0`?

Reply:
Removed this data member since it's used only once.